Run mypy job with Python 3.11 once it is supported

### Description

We should revert c3bfe6665e9d2420dfb2256c51e64ddc300f960a once mypy is sorted out (and if it's not mypy fault, then whatever we are doing wrong).

Maybe related to https://github.com/python/mypy/issues/12840, maybe we need to open a new issue over there?

I didn't save the run, but this was the error:

```
 [pytensor/link/c/exceptions.py]
 pytensor/link/c/exceptions.py:1: error: Module "setuptools._distutils.errors" has no attribute "CompileError"
```

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -61,7 +61,7 @@ repos:
         - numpy>=1.20
         - pandas
         - types-filelock
-        - types-setuptools
+        - types-setuptools>=67.6.0.8
         always_run: true
         require_serial: true
         pass_filenames: false

--- a/pytensor/link/c/exceptions.py
+++ b/pytensor/link/c/exceptions.py
@@ -1,4 +1,4 @@
-from setuptools._distutils.errors import CompileError as BaseCompileError
+from setuptools.errors import CompileError as BaseCompileError
 
 
 class MissingGXX(Exception):

